src/runtime/PythonEngine.cs

```diff
@@ -355,12 +355,26 @@ public static void Shutdown()
             {
                 return;
             }
+
+            // If we're embedded, we need to take the GIL, if it wasn't already 
+            // taken, as we may get called from the DomainUnload handler.
+            var pyGILState = PyGILState.PyGILState_UNLOCKED;
+            if (!Runtime.HostedInPython)
+            {
+                pyGILState = Runtime.PyGILState_Ensure();
+            }
+
```

Comment:
Since `Py.GIL()` no longer tries to force runtime initialization, I think the code would be easier to read with a simple `using (Py.GIL()) { ... }` around the `if`. I don't think it is necessary to check if we are hosted in Python

Reply:
Another issue is that the code below throws when Python error is set. And that `throw` prevents the GIL release code to be reached making the GIL unreleasable.

Reply:
I'll adjust it accordingly s.t. we can tag rc1.